File: base_conversion/base_conversion.java
package base_conversion;

class BaseConversion {

    private String baseConversion(int num, int n) {
        if (num == 0) {
            return "0";
        }

        String digits = "0123456789ABCDEF";
        StringBuilder sb = new StringBuilder();

        while (num != 0) {
            sb.append(digits.charAt(num % n));
            num /= n;
        }

        return sb.reverse().toString();
    }

    private int toDecimal(String num, int n) {
        int result = 0;
        String digits = "0123456789ABCDEF";

        for (int i = 0; i < num.length(); i++) {
            int digitValue = digits.indexOf(num.charAt(num.length() - 1 - i));
            result += digitValue * Math.pow(n, i);
        }

        return result;
    }

    public static void main(String[] args) {
        BaseConversion converter = new BaseConversion();

        System.out.println("===========================================");
        System.out.printf("%-10s %-10s %-10s\n", "Decimal", "Base", "Converted");
        System.out.println("===========================================");

        System.out.printf("%-10d %-10d %-10s\n", 3, 2, converter.baseConversion(3, 2));
        System.out.printf("%-10d %-10d %-10s\n", 10, 2, converter.baseConversion(10, 2));
        System.out.printf("%-10d %-10d %-10s\n", 166, 16, converter.baseConversion(166, 16));
        System.out.printf("%-10d %-10d %-10s\n", 166, 11, converter.baseConversion(166, 11));
        System.out.printf("%-10d %-10d %-10s\n", 100000, 16, converter.baseConversion(100000, 16));

        System.out.println("===========================================");
        System.out.printf("%-10s %-10s %-10s\n", "Value", "Base", "Decimal");
        System.out.println("===========================================");

        System.out.printf("%-10s %-10d %-10d\n", "11", 2, converter.toDecimal("11", 2));
        System.out.printf("%-10s %-10d %-10d\n", "1010", 2, converter.toDecimal("1010", 2));
        System.out.printf("%-10s %-10d %-10d\n", "A6", 16, converter.toDecimal("A6", 16));
        System.out.printf("%-10s %-10d %-10d\n", "142", 11, converter.toDecimal("142", 11));
        System.out.printf("%-10s %-10d %-10d\n", "186A0", 16, converter.toDecimal("186A0", 16));
        System.out.printf("%-10s %-10d %-10d\n", "0", 16, converter.toDecimal("0", 16));

        System.out.println("===========================================");
    }
}

File: base_conversion/base_conversion.cpp
#include <iostream>
#include <string>
#include <cmath>
#include <iomanip>
#include <cctype> 

class BaseConversion {
public:
    std::string baseConversion(int num, int base) {
        if (num == 0) {
            return "0";
        }

        const std::string digits = "0123456789ABCDEF";
        std::string result;

        while (num != 0) {
            result = digits[static_cast<size_t>(num % base)] + result;
            num /= base;
        }

        return result;
    }

    int toDecimal(const std::string& num, int base) {
        const std::string digits = "0123456789ABCDEF";
        int result = 0;

        for (std::string::size_type i = 0; i < num.size(); ++i) {
            char upperChar = static_cast<char>(toupper(static_cast<unsigned char>(num[num.size() - 1 - i])));
            size_t index = digits.find(upperChar);
            if (index == std::string::npos) {
                throw std::invalid_argument("Invalid character in input number.");
            }
            result += static_cast<int>(index) * static_cast<int>(pow(base, static_cast<int>(i)));
        }

        return result;
    }
};

int main() {
    BaseConversion converter;

    std::cout << "===========================================\n";
    std::cout << std::left << std::setw(10) << "Decimal" 
              << std::setw(10) << "Base" 
              << std::setw(10) << "Converted" << '\n';
    std::cout << "===========================================\n";

    std::cout << std::left << std::setw(10) << 3
              << std::setw(10) << 2
              << std::setw(10) << converter.baseConversion(3, 2) << '\n';

    std::cout << std::left << std::setw(10) << 10
              << std::setw(10) << 2
              << std::setw(10) << converter.baseConversion(10, 2) << '\n';

    std::cout << std::left << std::setw(10) << 166
              << std::setw(10) << 16
              << std::setw(10) << converter.baseConversion(166, 16) << '\n';

    std::cout << std::left << std::setw(10) << 166
              << std::setw(10) << 11
              << std::setw(10) << converter.baseConversion(166, 11) << '\n';

    std::cout << std::left << std::setw(10) << 100000
              << std::setw(10) << 16
              << std::setw(10) << converter.baseConversion(100000, 16) << '\n';

    std::cout << "===========================================\n";
    std::cout << std::left << std::setw(10) << "Value"
              << std::setw(10) << "Base"
              << std::setw(10) << "Decimal" << '\n';
    std::cout << "===========================================\n";

    std::cout << std::left << std::setw(10) << "11"
              << std::setw(10) << 2
              << std::setw(10) << converter.toDecimal("11", 2) << '\n';

    std::cout << std::left << std::setw(10) << "1010"
              << std::setw(10) << 2
              << std::setw(10) << converter.toDecimal("1010", 2) << '\n';

    std::cout << std::left << std::setw(10) << "A6"
              << std::setw(10) << 16
              << std::setw(10) << converter.toDecimal("A6", 16) << '\n';

    std::cout << std::left << std::setw(10) << "142"
              << std::setw(10) << 11
              << std::setw(10) << converter.toDecimal("142", 11) << '\n';

    std::cout << std::left << std::setw(10) << "186A0"
              << std::setw(10) << 16
              << std::setw(10) << converter.toDecimal("186A0", 16) << '\n';

    std::cout << std::left << std::setw(10) << "0"
              << std::setw(10) << 16
              << std::setw(10) << converter.toDecimal("0", 16) << '\n';

    std::cout << "===========================================\n";

    return 0;
}
